### DISCUSSION TITLE
Should https://validator.w3.org/ default to VNU?

### DISCUSSION THREAD

**Discussion by dd8** [validator/validator#1965] [General]
Followup to https://github.com/validator/validator/issues/719

Currently https://validator.w3.org/ directs to:

- the HTML5 VNU validator for documents with an HTML5 doctype or no doctype
- the legacy DTD validator for other doctypes (e.g. XHTML 1.0 and HTML 4.01)

## Pros:
- XHTML / HTML 4 sites that currently validate stay valid

## Cons:
- the DTD validator reports ARIA attributes like `role` and `aria-label` as errors, which is potentially harmful because it encourages authors to remove them
- the DTD validator treats new HTML5 self-closing elements like `track` as unclosed (so cause an error when added to an HTML 4 document to add subtitles)
- there are some unfixable bugs in the DTDs (because the DTDs are normative, but now obsolete) - see below
- there are lots of parsing differences between the DTD validator and the parsing algorithm implemented by browsers (and specified in the HTML Standard) 
- the OpenSP library used by the legacy DTD validator is unmaintained - the last code change and last release was 1.5.2 in 2007

## DTD Bugs
- **HTML 4.01** U+000C FORM FEED is reported as "non SGML character number 12" because it's forbidden in the DTD but is normatively specified as whitespace elsewhere in the spec: https://gist.github.com/dd8/8a8149c2ec7093dcf8caae6b9645ac0b#html-401
- **HTML 4.01** IDs are specified as case-sensitive in the normative text but case-insensitive in the DTDs. Happens because making element and attribute names case-insensitive in an SGML DTD also makes ID and and IDREFs case case-insensitive - it's an all-or-nothing option `NAMECASE GENERAL YES`. That means IDs and IDREFs that differ by case validate, but don't actually work.

These are unlikely to ever be fixed since the specs are marked as obsolete.

**Comment by dd8** [comment]
I've just come across an example where you might need a health warning on legacy DTD validation.

SharePoint.com sites use the following DOCTYPE
`<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">`

but they also use `<script nonce>` from HTML 5. 

Obviously they should change the DOCTYPE, but probably have some legacy compatibility reason that makes this difficult. 

Given removing `nonce` would break security I think it's worth adding a warning that removing attributes not defined in XHTML/HTML4 can cause problems (broken accessibility for ARIA attributes, broken security for nonce)

**Comment by j9t** [comment]
The central point is this, no, from #719?

> If someone is validating HTML 5 code via http://validator.w3.org/ (without the /nu) and miss out the DOCTYPE it defaults to validating HTML 4.01 Transitional and they get a bunch of HTML 4.01 messages.
>
> If instead a page with a missing DOCTYPE was routed to the VNU backend they'd get a single message:
> `Start tag seen without seeing a doctype first. Expected <!DOCTYPE html>`

I’d second that.

The description above doesn’t currently make clear what’s being suggested. But if it’s this, to go for living HTML if the document in question lacks a doctype, that seems reasonable to me.

**Reply by dd8** [reply]
https://github.com/validator/validator/issues/719 was fixed in 2018

The current behaviour of http://validator.w3.org  is:

1) no DOCTYPE - use HTML5 VNU validator (http://validator.w3.org/nu)
2) HTML 5 doctype like <!DOCTYPE html> - use HTML5 VNU validator
3) older doctypes (e.g. XHTML 1.0 and HTML 4.01) - uses the legacy DTD validator with normative DTDs for those DOCTYPEs (e.g. https://www.w3.org/TR/html4/strict.dtd)

The problem is no 3) - what should happen for old DOCTYPEs? Should the legacy DTD validator continue to be used or should every DOCTYPE use the VNU validator which knows about current attributes and elements supported by the browsers. 

The DTDs for older versions of HTML don't know anything about HTML elements and attributes introduced since 2008, so flag an error for any elements/attributes introduced in the last 17 years,

The problem is some of those are essential for security (e.g. the `nonce` and `sandbox` attributes) or accessibility (e.g. the `role` and `aria-*` attributes), so removing them to keep the DTD validator happy will break your site (as mentioned above some sites use new attributes with old DOCTYPEs)

For example the following code

```
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN"
            "http://www.w3.org/TR/html4/strict.dtd">
<html>
    <head>
    <style nonce='sha1-abced'>body{ color: white; }</style>
    </head>
    <body>
    </body>
</html>
```

reports this error on http://validator.w3.org

>  there is no attribute "NONCE"

At the very least there should be a health warning about removing unknown elements/attributes in legacy validation. The impact of removal varies by attribute:

- removing `nonce` from `<style>` or `<script>` would block the element due to Content-Security-Policy (this would probably be noticed quickly)
- removing `sandbox` from an `<iframe>` would not have any visible impact, but would be less secure (this might not be noticed until your site is compromised) 
- removing `role` or `aria-*` would likely cause problems for screen reader users
